### PR TITLE
feat(rum-react): add render/children prop support

### LINKED ISSUE #905: Support render props in ApmRoute component
+ Enhance the `elastic/apm-rum-react` package to add support for `render` props out of the box.

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ fix #905 
+ Added Unit/E2e tests for the same
+ Updated docs with the new changes and removed warning with `render` and `children` props. 

### Usage
```jsx
<ApmRoute path="/about" render={(props) => <AboutComponent {...props}/>} />
<ApmRoute path="/func" component={FunctionalComponent} />
<ApmRoute path="/topics" render={TopicComponent} />
<ApmRoute path="/topic/:id">{TopicComponent}</ApmRoute>
```

### CODE REVIEW

**Review by jahtalab — CHANGES_REQUESTED** [review]
Thanks @vigneshshanmugam 

[There are some logical differences](https://reactrouter.com/web/api/Route) between using `children`, `component` and `render` props with the `Route` component. At the moment we're always passing a component (regardless of the original prop). I think we should investigate if we can minimise the effect of these differences, or document them if we can't.

File: packages/rum-react/src/get-with-transaction.js [hunk 1]
```diff
@@ -54,9 +54,7 @@ function getWithTransaction(apm) {
 
       if (!Component) {
         const loggingService = apm.serviceFactory.getService('LoggingService')
-        loggingService.warn(
-          `${name} is not instrumented since component property is not provided`
-        )
+        loggingService.warn(`${name} is not instrumented by the agent`)
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-react/src/get-with-transaction.js:57]: Yes only if its null.

**Comment by vigneshshanmugam** [comment]
@jahtalab I went though the source code of the Route before proposing the change, There is no such such key difference that would kill the performance of the app under different circumstance, they are all the same under the hood (functions), For React it does not matter. 

Happy to add documentation. Let me know your thoughts.

**Comment by jahtalab** [comment]
@vigneshshanmugam , The differences are mentioned in their docs (linked above), more specifically:

> When you use component (instead of render or children, below) the router uses React.createElement to create a new React element from the given component. That means if you provide an inline function to the component prop, you would create a new component every render. This results in the existing component unmounting and the new component mounting instead of just updating the existing component. When using an inline function for inline rendering, use the render or the children prop (below)

This ^ seems to affect the performance, but more importantly it affects the functionality of the application, i.e. unmounting/mounting the component.

Also this part:

> Sometimes you need to render whether the path matches the location or not. In these cases, you can use the function children prop. It works exactly like render except that it gets called whether there is a match or not. The children render prop receives all the same route props as the component and render methods, except when a route fails to match the URL, then match is null. This allows you to dynamically adjust your UI based on whether or not the route matches

This also seems a difference in functionality depending on which prop is passed.

**Comment by jahtalab** [comment]
Meeting:
- We should add a test to verify the difference explain in the react-router docs
- We should add test from #924
- We should add to documents how it works in and what user might need to be aware of.

**Comment by ikalkov** [comment]
Any updates here?

**Pull request closed without merging**